src/realm/sync/noinst/client_reset.cpp

```diff
@@ -1799,17 +1826,142 @@ void client_reset::transfer_group(const Transaction& group_src, Transaction& gro
     }
 }
 
-client_reset::LocalVersionIDs client_reset::perform_client_reset_diff(DB& db_local, DBRef db_remote,
-                                                                      sync::SaltedFileIdent client_file_ident,
-                                                                      util::Logger& logger,
-                                                                      bool recover_local_changes)
+// A table without a "class_" prefix will not generate sync instructions.
+constexpr static std::string_view s_meta_reset_table_name("client_reset_metadata");
+constexpr static std::string_view s_pk_col_name("id");
+constexpr static std::string_view s_version_column_name("version");
+constexpr static std::string_view s_timestamp_col_name("event_time");
+constexpr static std::string_view s_reset_type_col_name("type_of_reset");
+constexpr int64_t metadata_version = 1;
+
+void remove_pending_client_resets(TransactionRef wt)
+{
+    REALM_ASSERT(wt);
+    if (auto table = wt->get_table(s_meta_reset_table_name)) {
+        if (table->size()) {
+            table->clear();
+        }
+    }
+}
+
+util::Optional<PendingReset> has_pending_reset(Transaction& rt)
+{
+    ConstTableRef table = rt.get_table(s_meta_reset_table_name);
+    if (!table || table->size() == 0) {
+        return util::none;
+    }
+    ColKey timestamp_col = table->get_column_key(s_timestamp_col_name);
+    ColKey type_col = table->get_column_key(s_reset_type_col_name);
+    ColKey version_col = table->get_column_key(s_version_column_name);
+    REALM_ASSERT(timestamp_col);
+    REALM_ASSERT(type_col);
+    REALM_ASSERT(version_col);
+    if (table->size() > 1) {
```

Comment:
wouldn't an assertion be more appropriate?

Reply:
Although it is an unexpected code path, there is real possibility of it being executed if the metadata format changes. In that case, I would rather have the reset soft fail and have the app continue to function rather than crash with an assertion failure.